### Battle of the Neighbourhoods - Toronto

In [4]:
import pandas as pd

# Question 1

#### Importing Data

In [7]:
import requests

In [19]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [15]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [20]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [21]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]


Dropping Borough which are not assigned

In [22]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]


Grouping the records based on Postal Code

In [23]:
df = df.groupby(['Postal Code']).head()
df


Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]


Checking for number of records where Neighbourhood is "Not assigned"

In [24]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [25]:
df = df.reset_index()
df


index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [26]:
df.drop(['index'], axis = 'columns', inplace = True)
df


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [27]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

## Question 2

#### Installing geocoder

In [28]:
pip install geocoder

     |████████████████████████████████| 98 kB 4.1 MB/s eta 0:00:011
Note: you may need to restart the kernel to use updated packages.


In [29]:
import geocoder


In [33]:
#Using link to a csv file that has the geographical coordinates of each postal code
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [32]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [34]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [35]:
data.dtypes


Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [44]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [45]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

## Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [51]:
from geopy.geocoders import Nominatim

In [47]:
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [62]:
!pip install folium
import folium

     |████████████████████████████████| 93 kB 4.0 MB/s  eta 0:00:01


In [64]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [83]:
CLIENT_ID = 'GW4NEDTJYXPWM5SX4VLFB2AMF13MDLDSUWDH4JUKEYYSE3U4' 
CLIENT_SECRET = '0GEQVUKUHMCABWS02ALNL3FN3IY0PMQASEJ13ZNYCDLT4BG4'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: GW4NEDTJYXPWM5SX4VLFB2AMF13MDLDSUWDH4JUKEYYSE3U4
CLIENT_SECRET:0GEQVUKUHMCABWS02ALNL3FN3IY0PMQASEJ13ZNYCDLT4BG4


Next, we create a function to get all the venue categories in Toronto

In [86]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [87]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [88]:
venues_in_toronto.shape

(1338, 5)

So we have 1317 records and 5 columns. Checking sample data

In [89]:
venues_in_toronto.head()


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood

In [90]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1322  Mimico NW, The Queensway West, South of Bloor,...   
1323  Mimico NW, The Queensway West, South of Bloor,...   
1324  Mimico NW, The Queensway West, South of Bloor,...   
1325  Mimico NW, The Queensway West, South of Bloor,...   
1326  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572             Tim Hortons   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1322               43.628841               -79.520999        South St. Burger   
1323               43.628841               -79.520999              Wingporium   
1324               43.628841               -79.520999               Dollarama   
1325               43.628841               -79.520999          Healthy Planet   
1326               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3               Coffee Shop  
4     Portuguese Restaurant  
...                     ...  
1322           Burger Joint  
1323            Wings Joint  
1324         Discount Store  
1325        Supplement Shop  
1326                 Bakery  

[416 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories

In [91]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.718518               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[240 rows x 4 columns]

There are around 232 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [93]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1333                  0        0                   0             0   
1334                  0        0                   0             0   
1335                  0        0                   0             0   
1336                  0        0                   0             0   
1337                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1333               0                0                 0                    0   
1334               0                0                 0                    0   
1335               0                0                 0                    0   
1336               0                0                 0                    0   
1337               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  Turkish Restaurant  \
0                0         0  ...              0                   0   
1                0         0  ...              0                   0   
2                0         0  ...              0                   0   
3                0         0  ...              0                   0   
4                0         0  ...              0                   0   
...            ...       ...  ...            ...                 ...   
1333             0         0  ...              0                   0   
1334             0         0  ...              0                   0   
1335             0         0  ...              0                   0   
1336             0         0  ...              0                   0   
1337             0         0  ...              0                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1333                              0                 0                      0   
1334                              0                 0                      0   
1335                              0                 0                      0   
1336                              0                 0                      0   
1337                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4  

Adding the neighbourhood to the encoded dataframe

In [94]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 241 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [96]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 241 columns]

Let's make a function to get the top most common venue categories

In [97]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [98]:
import numpy as np


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [99]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1              Pharmacy                Coffee Shop                   Pub   
2                  Bank                Pizza Place    Chinese Restaurant   
3                  Bank         Chinese Restaurant   Japanese Restaurant   
4        Sandwich Place         Italian Restaurant              Pharmacy   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0        Clothing Store          Dessert Shop  Eastern European Restaurant   
1        Sandwich Place                   Gym                      Gay Bar   
2           Bridal Shop         Shopping Mall               Sandwich Place   
3           Yoga Studio     Electronics Store  Eastern European Restaurant   
4       Thai Restaurant           Pizza Place                          Pub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Dumpling Restaurant             Drugstore             Donut Shop  
1            Donut Shop              Dive Bar    Distribution Center  
2                 Diner         Deli / Bodega             Restaurant  
3   Dumpling Restaurant             Drugstore             Donut Shop  
4            Restaurant                  Café                Butcher


Let's make the model to cluster our Neighbourhoods

In [100]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [101]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans


KMeans(n_clusters=5, random_state=0)

In [102]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 2, 1, 0, 4, 0, 1, 0, 0, 0, 3, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0], dtype=int32)

In [103]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [104]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0                  Park   
1  43.725882 -79.315572             0.0          Intersection   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio      Department Store   
1          Hockey Arena           Pizza Place     French Restaurant   
2                  Park                Bakery        Breakfast Spot   
3     Accessories Store              Boutique             Gift Shop   
4           Yoga Studio                 Diner        Sandwich Place   

         5th Most Common Venue  6th Most Common Venue  7th Most Common Venue  \
0  Eastern European Restaurant    Dumpling Restaurant              Drugstore   
1                  Coffee Shop  Portuguese Restaurant       Department Store   
2                      Theater      French Restaurant  Performing Arts Venue   
3       Furniture / Home Store            Event Space            Coffee Shop   
4        Portuguese Restaurant                   Park     Mexican Restaurant   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0            Donut Shop                Dog Run               Dive Bar  
1          Dessert Shop     Dim Sum Restaurant                  Diner  
2        Chocolate Shop                    Pub             Restaurant  
3         Women's Store  Vietnamese Restaurant       Airport Terminal  
4    Italian Restaurant             Hobby Shop    Fried Chicken Joint

In [105]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [106]:

import matplotlib.cm as cm
import matplotlib.colors as colors

In [107]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [108]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             0.0          Intersection   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
7          North York             0.0                   Gym   
..                ...             ...                   ...   
97   Downtown Toronto             0.0                  Café   
98          Etobicoke             0.0                  Pool   
99   Downtown Toronto             0.0           Coffee Shop   
100      East Toronto             0.0      Recording Studio   
102         Etobicoke             0.0                   Gym   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1            Hockey Arena           Pizza Place     French Restaurant   
2                    Park                Bakery        Breakfast Spot   
3       Accessories Store              Boutique             Gift Shop   
4             Yoga Studio                 Diner        Sandwich Place   
7              Beer Store   Japanese Restaurant           Coffee Shop   
..                    ...                   ...                   ...   
97            Coffee Shop            Restaurant    Seafood Restaurant   
98                  River           Yoga Studio         Deli / Bodega   
99               Beer Bar        Breakfast Spot       Bubble Tea Shop   
100         Auto Workshop            Skate Park                  Park   
102           Social Club        Discount Store                Bakery   

           5th Most Common Venue  6th Most Common Venue  \
1                    Coffee Shop  Portuguese Restaurant   
2                        Theater      French Restaurant   
3         Furniture / Home Store            Event Space   
4          Portuguese Restaurant                   Park   
7             Dim Sum Restaurant         Discount Store   
..                           ...                    ...   
97                           Gym               Tea Room   
98   Eastern European Restaurant    Dumpling Restaurant   
99                  Burger Joint     Salon / Barbershop   
100           Light Rail Station          Burrito Place   
102                Tanning Salon          Burrito Place   

     7th Most Common Venue 8th Most Common Venue   9th Most Common Venue  \
1         Department Store          Dessert Shop      Dim Sum Restaurant   
2    Performing Arts Venue        Chocolate Shop                     Pub   
3              Coffee Shop         Women's Store   Vietnamese Restaurant   
4       Mexican Restaurant    Italian Restaurant              Hobby Shop   
7              Supermarket  Caribbean Restaurant                    Café   
..                     ...                   ...                     ...   
97               Gastropub  Gym / Fitness Center                   Hotel   
98               Drugstore            Donut Shop                 Dog Run   
99                    Café            Restaurant        Ramen Restaurant   
100         Farmers Market  Fast Food Restaurant                 Butcher   
102           Burger Joint  Fast Food Restaurant  Thrift / Vintage Store   

    10th Most Common Venue  
1                    Diner  
2               Restaurant  
3         Airport Terminal  
4      Fried Chicken Joint  
7       Chinese Restaurant  
..                     ...  
97            Concert Hall  
98                Dive Bar  
99                     Pub  
100             Restaurant  
102             Kids Store  

[86 rows x 12 columns]

Cluster 2

In [110]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             1.0                  Park   
21              York             1.0                  Park   
35         East York             1.0                  Park   
49        North York             1.0                  Park   
61   Central Toronto             1.0                  Park   
64              York             1.0                  Park   
66        North York             1.0                  Park   
85       Scarborough             1.0            Playground   
91  Downtown Toronto             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio      Department Store   
21               Women's Store                  Pool         Deli / Bodega   
35           Convenience Store          Intersection           Yoga Studio   
49  Construction & Landscaping                Bakery           Yoga Studio   
61                 Swim School              Bus Line           Yoga Studio   
64                 Yoga Studio      Department Store     Electronics Store   
66           Convenience Store           Yoga Studio      Department Store   
85                        Park                Bakery          Intersection   
91                Tennis Court                 Trail            Playground   

          5th Most Common Venue        6th Most Common Venue  \
0   Eastern European Restaurant          Dumpling Restaurant   
21  Eastern European Restaurant          Dumpling Restaurant   
35                 Dessert Shop            Electronics Store   
49                 Dessert Shop            Electronics Store   
61                 Dessert Shop            Electronics Store   
64  Eastern European Restaurant          Dumpling Restaurant   
66            Electronics Store  Eastern European Restaurant   
85             Department Store  Eastern European Restaurant   
91                 Dessert Shop           Dim Sum Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                     Drugstore            Donut Shop               Dog Run   
21                    Drugstore            Donut Shop               Dog Run   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
49  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
61  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
64                    Drugstore            Donut Shop               Dog Run   
66          Dumpling Restaurant             Drugstore            Donut Shop   
85          Dumpling Restaurant             Drugstore            Donut Shop   
91                        Diner        Discount Store   Distribution Center   

   10th Most Common Venue  
0                Dive Bar  
21               Dive Bar  
35             Donut Shop  
49             Donut Shop  
61             Donut Shop  
64               Dive Bar  
66                Dog Run  
85                Dog Run  
91                Dog Run

Cluster 3

In [112]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6   Dumpling Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run               Dive Bar

Cluster 4

In [113]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  Cluster Labels       1st Most Common Venue  \
57   North York             3.0              Baseball Field   
101   Etobicoke             3.0  Construction & Landscaping   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57            Yoga Studio           Event Space           Escape Room   
101        Baseball Field           Yoga Studio    Dim Sum Restaurant   

    5th Most Common Venue        6th Most Common Venue  \
57      Electronics Store  Eastern European Restaurant   
101           Escape Room            Electronics Store   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57           Dumpling Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

    10th Most Common Venue  
57                 Dog Run  
101             Donut Shop

Cluster 5